src/middlewared/middlewared/plugins/service.py

```diff
@@ -74,16 +71,43 @@ def result(task):
                     self.logger.warning('Task %r failed', exc_info=True)
 
             if result is None:
-                entry['state'] = 'UNKNOWN'
-                entry['pids'] = []
-            else:
-                entry['state'] = 'RUNNING' if result.running else 'STOPPED'
-                entry['pids'] = result.pids
+                return {
+                    'service': entry['service'],
+                    'state': 'UNKNOWN',
+                    'pids': []
+                }
+
+            return {
+                'service': entry['service'],
+                'state': 'RUNNING' if result.running else 'STOPPED',
+                'pids': result.pids
+            }
+
+        return {srv['service']: srv for srv in map(result, jobs)}
+
+    @private
+    async def service_extend(self, svc, ctx):
+        return svc | ctx.get(svc['service'], {'state': 'UNKNOWN', 'pids': []})
+
+    @filterable
+    async def query(self, filters, options):
+        """
+        Query all system services with `query-filters` and `query-options`.
+
+        Supports the following extra options:
+        `skip_state` - performance optimization to avoid getting service state.
+        """
+        default_options = {
+            'prefix': self._config.datastore_prefix,
+            'extend': self._config.datastore_extend,
+            'extend_context': self._config.datastore_extend_context
+        }
 
-            return entry
+        if set(filter_getattrs(filters)) & {'state', 'pids'}:
+            services = await self.middleware.call('datastore.query', 'services.services', [], default_options)
```

Comment:
Why do you discard other options (e.g. `get`) and why is this a special case?

Reply:
They're not discarded. They're picked up in the next line.

Reply:
```
root@carl[/mnt/dozer/middleware/src/middlewared]# midclt call service.query '[["state", "=", "RUNNING"]]' '{"get": true}' | jq
{
  "id": 4,
  "service": "cifs",
  "enable": true,
  "state": "RUNNING",
  "pids": [
    533202
  ]
}
```
Seems like `get` is working as expected in this case. How can we apply `options` correctly if `filters` are not passed to datastore? Filters are not passed to datastore plugin because they fail with KeyError.

Reply:
Still will not contain `get` or other options and I am still don't understand why is this a special case worth special handling.